### PR TITLE
Fix New-Item -Path/-Target with wildcard char

### PR DESCRIPTION
**Pull request by kwkam**
Reopen #7404

# PR Summary

Unescape non-literal path when -Name is not set.
Fix failure with symlink/junction target which contains special character.

## PR Context

This works around for New-Item treating -Path as literal path while
it can also be globbable. For example, assuming there is ```"[file]1"```
in current directory, and tab completion suggests ```'./`[file`]1'```
for the -Path argument, but
```pwsh
New-Item -Path './`[file`]2'
```
will create a file named ```"`[file`]2"``` instead of ```"[file]2"```.

Also fixes an issue that New-Item does not handle -Target properly
when it contains escaped wildcard character. For example, there is
a directory ```"[dir]1"``` in current directory, and tab completion
suggests ```'./`[dir`]1'``` for the -Target argument, but
```pwsh
New-Item -Type Junction -Path './`[dir`]2' -Target './`[dir`]1'
```
will complain that it cannot find the item.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1 [hunk 1]
```diff
@@ -103,6 +104,12 @@ Describe "New-Item" -Tags "CI" {
         Test-Path $FullyQualifiedFile | Should -BeTrue
     }
 
+    It "Should create a file with correct name when Name switch is not used and Path contains special char" {
+        New-Item -Path $FullyQualifiedFileSp -ItemType file
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]:
```suggestion
        New-Item -Path $FullyQualifiedFileSp -ItemType file > $null
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1 [hunk 2]
```diff
@@ -103,6 +104,12 @@ Describe "New-Item" -Tags "CI" {
...
+
+        $FullyQualifiedFileSp | Should -Exist
+    }
+
     It "Should be able to create a multiple items in different directories" {
         $FullyQualifiedFile2 = Join-Path -Path $tmpDirectory -ChildPath test2.txt
         New-Item -ItemType file -Path $FullyQualifiedFile, $FullyQualifiedFile2
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]:
```suggestion
        New-Item -ItemType file -Path $FullyQualifiedFile, $FullyQualifiedFile2 > $null
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1 [hunk 3]
```diff
@@ -249,6 +262,21 @@ Describe "New-Item with links" -Tags @('CI', 'RequireAdminOnWindows') {
         Test-Path $FullyQualifiedLink | Should -BeFalse
     }
 
+    It "Should create symbolic link with name contains special char" {
+        $null = New-Item -Path $tmpDirectory -Name $testlinkSrcSpName -ItemType File
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]:
```suggestion
        New-Item -Path $tmpDirectory -Name $testlinkSrcSpName -ItemType File > $null
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1 [hunk 4]
```diff
@@ -249,6 +262,21 @@ Describe "New-Item with links" -Tags @('CI', 'RequireAdminOnWindows') {
...
+        $FullyQualifiedLSrcSp | Should -Exist
+
+        $null = New-Item -Path $FullyQualifiedLinkSp -Target $FullyQualifiedLSrcSp -ItemType SymbolicLink
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]:
```suggestion
        New-Item -Path $FullyQualifiedLinkSp -Target $FullyQualifiedLSrcSp -ItemType SymbolicLink > $null
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1 [hunk 5]
```diff
@@ -109,6 +112,14 @@ Describe "Remove-Item" -Tags "CI" {
 	    Test-Path (Join-Path -Path $testpath -ChildPath file1.wav) | Should -BeFalse
 	    Test-Path (Join-Path -Path $testpath -ChildPath file2.wav) | Should -BeFalse
 	}
+
+        It "Should be able to remove file when path contains special char" {
+            New-Item -Path $testpath -Name $testfileSpName -ItemType File -Force
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]:
```suggestion
            New-Item -Path $testpath -Name $testfileSpName -ItemType File -Force > $null
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1 [hunk 6]
```diff
@@ -109,6 +112,14 @@ Describe "Remove-Item" -Tags "CI" {
...
+            $testfilepathSp | Should -Exist
+
+            Remove-Item -Path $testfilepathSp
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Remove-Item.Tests.ps1]:
```suggestion
            Remove-Item -Path $testfilepathSp -Force
```

File: test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1 [hunk 7]
```diff
@@ -249,6 +262,21 @@ Describe "New-Item with links" -Tags @('CI', 'RequireAdminOnWindows') {
...
+        New-Item -Path $tmpDirectory -Name $testlinkSrcSpName -ItemType File > $null
+        $FullyQualifiedLSrcSp | Should -Exist
+
+        New-Item -Path $FullyQualifiedLinkSp -Target $FullyQualifiedLSrcSp -ItemType SymbolicLink > $null
+        $FullyQualifiedLinkSp | Should -Exist
+
+        $expectedTarget = Join-Path -Path $tmpDirectory -ChildPath $testlinkSrcSpName
+
+        $fileInfo = Get-ChildItem -Path $FullyQualifiedLinkSp
+        $fileInfo.Target | Should -Match ([regex]::Escape($expectedTarget))
```
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]: Why can't we do an -BeExactly assertion here instead of a -Match?

Comment (kwkam) [reply test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1]: I guess we can.
Actually this was created using the tests in the same file as reference:
https://github.com/PowerShell/PowerShell/blob/2737e74d864de909c7a359fad31e75e2cf0db34b/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1#L217
https://github.com/PowerShell/PowerShell/blob/2737e74d864de909c7a359fad31e75e2cf0db34b/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1#L242
Perhaps we have to review the New-Item test sometime.

File: src/System.Management.Automation/engine/SessionStateContainer.cs [hunk 8]
```diff
@@ -3535,9 +3535,13 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
                             throw PSTraceSource.NewInvalidOperationException(SessionStateStrings.PathNotFound, targetPath);
                         }
 
-                        // If the original target was a relative path, we want to leave it as relative if it did not require
-                        // globbing to resolve.
-                        if (WildcardPattern.ContainsWildcardCharacters(targetPath))
+                        // If the original target was a relative path, we want to leave it as relative
+                        if (targetPath.StartsWith(".", StringComparison.OrdinalIgnoreCase))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateContainer.cs]:
```suggestion
                        if (targetPath.StartsWith('.'))
```

**Review by SteveL-MSFT — APPROVED** [review]

File: src/System.Management.Automation/engine/SessionStateContainer.cs [hunk 9]
```diff
@@ -3536,10 +3536,11 @@ internal Collection<PSObject> NewItem(string[] paths, string name, string type,
                         }
 
                         // If the original target was a relative path, we want to leave it as relative
-                        if (targetPath.StartsWith(".", StringComparison.OrdinalIgnoreCase))
+                        if (targetPath.StartsWith("."))
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/SessionStateContainer.cs]: Please use char '.'. Below too.

**Comment by adityapatwardhan** [comment]
I believe we should not change this behavior. To workaround your problem you could simply use the following:

```PowerShell
PS D:\temp\newitem> New-Item -Name `[file`]2


    Directory: D:\temp\newitem

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----         5/10/2019 10:32 AM              0 [file]2
```

**Comment by kwkam** [comment]
The problem is the inconsistency of `-Path` behaviour in the `*-Item` cmdlets.
```PowerShell
# to get '[file]1'
Get-Item -Path './`[file`]1'

# to move '[file]1'
Move-Item -Path './`[file`]1' -Destination ./dir

# to copy '[file]1'
Copy-Item -Path './dir/`[file`]1' -Destination ..

# to remove '[file]1'
Remove-Item -Path './dir/`[file`]1'
```
All above commands behave correctly with tab completion, but when we do
```PowerShell
# what if we try to overwrite '[file]1'?
New-Item -Path './`[file`]1' -Value 'This is [file]1'
```
A new file ``` `[file`]1 ``` is created instead, which is unexpected, and in my opinion, is a flaw in the cmdlet since `-Path` is supposed to handle wildcard path and escaped path.

Of course, we can use `-Name` to workaround the issue (because there is no `-LiteralPath` in `New-Item`), but I believe `-Name` is not the proper replacement, as @mklement0 commented in https://github.com/PowerShell/PowerShell/issues/6232#issuecomment-368394604, `-Name` is not supposed to be path. Anyway, even if we have the workaround, a bug needs a fix.

**Comment by iSazonov** [comment]
@adityapatwardhan Could you please update your review?

**Comment by adityapatwardhan** [comment]
Changing the behavior of `Path` is a breaking change and hence I would like the committee to review it.

**Comment by joeyaiello** [comment]
@PowerShell/powershell-committee discussed this one. 

The Issue at hand is that tab completion uses single quotes and escaped characterss in what amounts to a literal path, and you only need one or the other to work properly. We agreed not to fix it in the `New-Item` behavior, which would be a significant breaking change, but we should fix it in the tab completion by not adding backticks (because double quotes will break where `$` is in the filename). 

We'd welcome a PR to fix the tab completion, but given that we're not taking this one, closing it out.

**Comment by iSazonov** [comment]
@kwkam Have you plans to fix the tab completion?

**Pull request closed without merging**